lib/pbench/agent/results.py

```diff
@@ -123,60 +206,143 @@ def make_result_tb(self) -> str:
+                        f"Failed to create tar ball; 'xz' return code: {xz_proc.returncode:d}",
                     )
+                    raise RuntimeError(msg)
+                else:
+                    try:
+                        e_file.unlink()
+                    except FileNotFoundError:
+                        pass
+                    except Exception as unlink_exc:
+                        self.logger.warning(
+                            "Failed to remove 'tar' stderr file, %s: '%s'",
+                            e_file,
+                            unlink_exc,
+                        )
+            else:
+                # We explicitly ignore the return code from the optional 'xz' process.
+                msg = self._unlink_tarball(
+                    tarball,
+                    f"Failed to create tar ball; 'tar' return code: {tar_proc.returncode:d}",
+                )
+                raise RuntimeError(msg)
+        try:
+            (tar_len, tar_md5) = md5sum(tarball)
+        except Exception:
+            msg = self._unlink_tarball(
+                tarball,
+                f"Failed to verify and generate MD5 for created tar ball, '{tarball}'",
+            )
+            raise RuntimeError(msg)
 
         # The contract with the caller is to just return the full path to the
-        # created tarball.
-        return str(tarball)
+        # created tar ball.
+        return TarballRecord(tarball, tar_len, tar_md5)
 
 
 class CopyResultTb:
-    """CopyResultTb - Use the server's HTTP PUT method to upload a tarball
+    """Interfaces for copying result tar balls remotely using the server's HTTP
+    PUT method for uploads.
     """
 
+    class FileUploadError(Exception):
+        """Raised when the uploading of file to server has failed"""
+
+        pass
+
     def __init__(
-        self, controller: str, tarball: str, config: PbenchAgentConfig, logger: Logger
+        self,
+        controller: str,
+        tarball: str,
+        tarball_len: int,
+        tarball_md5: str,
+        config: PbenchAgentConfig,
+        logger: Logger,
     ):
-        """CopyResultTb contructor - raises FileNotFoundError if the given
-        tar ball does not exist, and a ValueError if the given controller is
-        not a valid hostname.
+        """Contructor for object representing tar ball to be copied remotely.
+
+        Raises
+            ValueError          if the given controller is not a valid hostname
+            FileNotFoundError   if the given tar ball does not exist
         """
-        if validate_hostname(controller) != 0:
-            raise ValueError(f"Controller {controller!r} is not a valid host name")
         self.controller = controller
         self.tarball = Path(tarball)
         if not self.tarball.exists():
-            raise FileNotFoundError(f"Tarball '{self.tarball}' does not exist")
+            raise FileNotFoundError(f"Tar ball '{self.tarball}' does not exist")
```

Comment:
It's a trivial thing, but it seems like we should do this check before assigning `self.tarball`, indeed perhaps before doing any work....

Reply:
Yes, but why create the `Path()` object in another variable to do the test before the assignment?